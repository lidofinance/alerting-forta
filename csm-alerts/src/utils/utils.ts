import { Contract, EventFilter, Event } from 'ethers'

const LOG_FILTER_CHUNK = 2000

export async function getLogsByChunks(contract: Contract, filter: EventFilter, startblock: number, endBlock: number) {
  const events: Event[] = []
  let endBlockChunk
  let startBlockChunk = startblock
  do {
    endBlockChunk =
      endBlock > startBlockChunk + LOG_FILTER_CHUNK - 1 ? startBlockChunk + LOG_FILTER_CHUNK - 1 : endBlock
    const eventsChunk = await contract.queryFilter(filter, startBlockChunk, endBlockChunk)
    events.push(...eventsChunk)
    startBlockChunk = endBlockChunk + 1
  } while (endBlockChunk < endBlock)
  return events
}
