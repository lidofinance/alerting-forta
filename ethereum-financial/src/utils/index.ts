import BigNumber from "bignumber.js";

const TEN_TO_18 = new BigNumber(10).pow(18);

export function formatEth(amount: any, dp: number): string {
  return new BigNumber(String(amount)).div(TEN_TO_18).toFixed(dp);
}

export function formatDelay(fullDelaySec: number) {
  let sign = fullDelaySec >= 0 ? 1 : -1;
  let delayHours = 0;
  let delayMin = Math.floor((sign * fullDelaySec) / 60);
  let delaySec = sign * fullDelaySec - delayMin * 60;
  if (delayMin >= 60) {
    delayHours = Math.floor(delayMin / 60);
    delayMin -= delayHours * 60;
  }
  return (
    (sign == 1 ? "" : "-") +
    (delayHours > 0 ? `${delayHours} hrs ` : "") +
    (delayMin > 0 ? `${delayMin} min ` : "") +
    `${delaySec} sec`
  );
}
