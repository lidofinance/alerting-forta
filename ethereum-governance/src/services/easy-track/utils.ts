import { SAFES } from 'constants/easy-track'
import { STONKS } from 'constants/stonks'
import { getSafeAddress, etherscanAddress } from '../../shared/string'
import { Blockchain } from '../../shared/contracts'

export const getMotionCreatorNamedLink = (address: string) => {
  const ethSafes = SAFES.Ethereum
  const safeName = ethSafes.find((safeData: string[]) => safeData[0].toLowerCase() === address.toLowerCase())
  if (safeName) {
    return getSafeAddress(address, safeName[1])
  } else {
    return etherscanAddress(address)
  }
}

export const getStonksContractInfo = (address: string) => {
  return STONKS.find((c) => c.address === address)
}

export const getSafeNameByAddress = (address: string) => {
  let safeName = address
  Object.keys(SAFES).forEach((key) => {
    const safe = SAFES[key as Blockchain].find(
      (safeData: string[]) => safeData[0].toLowerCase() === address.toLowerCase(),
    )
    if (safe) {
      safeName = safe[1]
      return
    }
  })
  return safeName
}
