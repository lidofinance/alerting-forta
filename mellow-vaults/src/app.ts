import { getEthersProvider, fetchJwt, Finding, verifyJwt } from 'forta-agent'
import { LIDO_STETH_ADDRESS } from 'constants/common'
import { ETHProvider } from './clients/eth_provider'
import { FormatterWithEIP1898 } from './clients/eth_formatter'
import { DataRW } from './shared/mutex'
import * as Winston from 'winston'
import * as E from 'fp-ts/Either'
import { BorderTime, HealthChecker, MaxNumberErrorsPerBorderTime } from './services/health-checker/HealthChecker.srv'
import { LidoDAO__factory } from './generated'
import { VaultWatcherSrv } from './services/vault-watcher/VaultWatcher.srv'
import { MultisigWatcherSrv } from './services/multisig-watcher/MultisigWatcher.srv'

export type Container = {
  ethClient: ETHProvider
  VaultWatcherSrv: VaultWatcherSrv
  MultisigWatcherSrv: MultisigWatcherSrv
  findingsRW: DataRW<Finding>
  healthChecker: HealthChecker
}

export class App {
  private static instance: Container

  public static async getJwt(): Promise<E.Either<Error, string>> {
    let token: string
    try {
      token = await fetchJwt({})
    } catch (e) {
      return E.left(new Error(`Could not fetch jwt. Cause ${e}`))
    }

    if (process.env.NODE_ENV === 'production') {
      try {
        const isTokenOk = await verifyJwt(token)
        if (!isTokenOk) {
          return E.left(new Error(`Token verification failed`))
        }
      } catch (e) {
        return E.left(new Error(`Token verification failed`))
      }
    }

    return E.right(token)
  }

  public static async getInstance(): Promise<Container> {
    if (!App.instance) {
      const ethersProvider = getEthersProvider()
      ethersProvider.formatter = new FormatterWithEIP1898()

      const stethContract = LidoDAO__factory.connect(LIDO_STETH_ADDRESS, ethersProvider)

      const ethClient = new ETHProvider(ethersProvider, stethContract)

      const logger: Winston.Logger = Winston.createLogger({
        format: Winston.format.simple(),
        transports: [new Winston.transports.Console()],
      })

      const vaultWatcherSrv = new VaultWatcherSrv(logger, ethClient)
      const multisigWatcherSrv = new MultisigWatcherSrv(logger)

      App.instance = {
        ethClient: ethClient,
        VaultWatcherSrv: vaultWatcherSrv,
        MultisigWatcherSrv: multisigWatcherSrv,
        findingsRW: new DataRW([]),
        healthChecker: new HealthChecker(BorderTime, MaxNumberErrorsPerBorderTime),
      }
    }

    return App.instance
  }
}
