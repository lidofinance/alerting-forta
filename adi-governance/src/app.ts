import { fetchJwt, Finding, getEthersProvider, verifyJwt } from 'forta-agent'
import { ETHProvider } from './clients/eth_provider'
import { DataRW } from './utils/mutex'
import * as Winston from 'winston'
import * as E from 'fp-ts/Either'
import { BorderTime, HealthChecker, MaxNumberErrorsPerBorderTime } from './services/health-checker/health-checker.srv'
import { BnbAdiSrv } from './services/bnb-adi/BnbAdi.srv'
import { CrossChainController__factory } from './generated'
import { ADI_CROSS_CHAIN_CONTROLLER } from './utils/constants'

export type Container = {
  ethClient: ETHProvider
  BnbAdiSrv: BnbAdiSrv
  findingsRW: DataRW<Finding>
  healthChecker: HealthChecker
}

export class App {
  private static instance: Container

  public static async getJwt(): Promise<E.Either<Error, string>> {
    let token: string
    try {
      token = await fetchJwt({})
    } catch (e) {
      return E.left(new Error(`Could not fetch jwt. Cause ${e}`))
    }

    if (process.env.NODE_ENV === 'production') {
      try {
        const isTokenOk = await verifyJwt(token)
        if (!isTokenOk) {
          return E.left(new Error(`Token verification failed`))
        }
      } catch (e) {
        return E.left(new Error(`Token verification failed`))
      }
    }

    return E.right(token)
  }

  public static async getInstance(): Promise<Container> {
    if (!App.instance) {
      const ethersProvider = getEthersProvider()

      const crossChainControllerContract = CrossChainController__factory.connect(
        ADI_CROSS_CHAIN_CONTROLLER,
        ethersProvider,
      )

      const ethClient = new ETHProvider(ethersProvider, crossChainControllerContract)

      const logger: Winston.Logger = Winston.createLogger({
        format: Winston.format.simple(),
        transports: [new Winston.transports.Console()],
      })

      const bnbAdiSrv = new BnbAdiSrv(logger, ethClient)

      App.instance = {
        ethClient: ethClient,
        BnbAdiSrv: bnbAdiSrv,
        findingsRW: new DataRW([]),
        healthChecker: new HealthChecker(BorderTime, MaxNumberErrorsPerBorderTime),
      }
    }

    return App.instance
  }
}
