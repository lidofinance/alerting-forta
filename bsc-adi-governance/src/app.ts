import { Finding, getEthersProvider } from 'forta-agent'
import * as Winston from 'winston'
import { BSCProvider } from './clients/bsc_provider'
import { EventWatcherSrv } from './services/event-watcher/EventWatcher.srv'
import { CrossChainControllerSrv } from './services/cross-chain-controller/CrossChainController.srv'
import { AclChangesSrv } from './services/acl-changes/AclChanges.srv'
import { DataRW } from './utils/mutex'
import { BorderTime, HealthChecker, MaxNumberErrorsPerBorderTime } from './services/health-checker/health-checker.srv'
import { CROSS_CHAIN_CONTROLLER_ADDRESS, CROSS_CHAIN_EXECUTOR_ADDRESS } from './utils/constants'
import { CrossChainController__factory } from './generated'
import { getCrossChainExecutorEvents } from './utils/events/cross_chain_executor_events'

export type Container = {
  bscClient: BSCProvider
  crossChainControllerSrv: CrossChainControllerSrv
  crossChainExecutorWatcherSrv: EventWatcherSrv
  bscAclChangesSrv: AclChangesSrv
  findingsRW: DataRW<Finding>
  healthChecker: HealthChecker
}

export class App {
  private static instance: Container

  public static async getInstance(): Promise<Container> {
    if (!App.instance) {
      const ethersProvider = getEthersProvider()

      const crossChainControllerRunner = CrossChainController__factory.connect(
        CROSS_CHAIN_CONTROLLER_ADDRESS,
        ethersProvider,
      )

      const bscClient = new BSCProvider(ethersProvider, crossChainControllerRunner)

      const logger: Winston.Logger = Winston.createLogger({
        format: Winston.format.simple(),
        transports: [new Winston.transports.Console()],
      })

      const crossChainControllerSrv = new CrossChainControllerSrv(logger, bscClient, CROSS_CHAIN_CONTROLLER_ADDRESS)
      const crossChainExecutorWatcherSrv = new EventWatcherSrv(
        'CrossChainExecutorWatcher',
        getCrossChainExecutorEvents(CROSS_CHAIN_EXECUTOR_ADDRESS),
        logger,
      )
      const bscAclChangesSrv = new AclChangesSrv(logger, bscClient)

      App.instance = {
        bscClient,
        crossChainControllerSrv,
        crossChainExecutorWatcherSrv,
        bscAclChangesSrv,
        findingsRW: new DataRW([]),
        healthChecker: new HealthChecker(BorderTime, MaxNumberErrorsPerBorderTime),
      }
    }

    return App.instance
  }
}
