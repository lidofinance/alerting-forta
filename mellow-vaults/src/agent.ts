import {
  BlockEvent,
  Finding,
  FindingSeverity,
  FindingType,
  HandleBlock,
  HealthCheck,
  HandleTransaction,
} from 'forta-agent'
import * as process from 'process'
import { argv } from 'process'
import { InitializeResponse } from 'forta-agent/dist/sdk/initialize.response'
import { Initialize } from 'forta-agent/dist/sdk/handlers'
import * as E from 'fp-ts/Either'
import { App } from './app'
import { elapsedTime } from './shared/time'
import Version from './shared/version'
import { Metadata } from './entity/metadata'
import { TransactionEvent } from 'forta-agent'
import { BlockDto } from './entity/events'

export function initialize(): Initialize {
  const metadata: Metadata = {
    'version.commitHash': Version.commitHash,
    'version.commitMsg': Version.commitMsg,
  }

  return async function (): Promise<InitializeResponse | void> {
    const startTime = new Date().getTime()
    const app = await App.getInstance()

    const latestBlock = await app.ethClient.getStartedBlockForApp(argv)
    if (E.isLeft(latestBlock)) {
      console.error(`Error: ${latestBlock.left.message}`)
      console.error(`Stack: ${latestBlock.left.stack}`)

      process.exit(1)
    }

    const lastBlockNumber = latestBlock.right.number
    const [vaultWatcherSrvErr] = await Promise.all([
      app.VaultWatcherSrv.initialize(lastBlockNumber),
      app.MultisigWatcherSrv.initialize(lastBlockNumber),
      app.AclChangesSrv.initialize(lastBlockNumber),
    ])

    if (vaultWatcherSrvErr !== null) {
      console.error(`Error: ${vaultWatcherSrvErr.message}`)
      console.error(`Stack: ${vaultWatcherSrvErr.stack}`)

      process.exit(1)
    }

    const agents = [app.VaultWatcherSrv.getName(), app.MultisigWatcherSrv.getName(), app.AclChangesSrv.getName()]
    metadata.agents = '[' + agents.toString() + ']'

    await app.findingsRW.write([
      Finding.fromObject({
        name: `Agent launched`,
        description: `Version: ${Version.desc}`,
        alertId: 'LIDO-AGENT-LAUNCHED',
        severity: FindingSeverity.Info,
        type: FindingType.Info,
        metadata,
      }),
    ])

    console.log(elapsedTime('Agent.initialize', startTime) + '\n')
  }
}

export const handleBlock = (): HandleBlock => {
  return async function (blockEvent: BlockEvent): Promise<Finding[]> {
    console.log(`#ETH block: ${blockEvent.block.number}`)
    const startTime = new Date().getTime()

    const app = await App.getInstance()

    const findings: Finding[] = []
    const findingsAsync = await app.findingsRW.read()
    if (findingsAsync.length > 0) {
      findings.push(...findingsAsync)
    }

    const blockDtoEvent: BlockDto = {
      number: blockEvent?.block?.number,
      timestamp: blockEvent?.block?.timestamp,
      parentHash: blockEvent?.block?.parentHash,
    }

    const vaultWatcherSrvFindings = await app.VaultWatcherSrv.handleBlock(blockDtoEvent)
    const aclChangesSrvFindings = await app.AclChangesSrv.handleBlock(blockDtoEvent)
    findings.push(...vaultWatcherSrvFindings, ...aclChangesSrvFindings)

    console.log(elapsedTime('handleBlock', startTime) + '\n')
    return findings
  }
}

export const handleTransaction = (): HandleTransaction => {
  return async function (txEvent: TransactionEvent): Promise<Finding[]> {
    const app = await App.getInstance()

    const vaultFindings = await app.VaultWatcherSrv.handleTransaction(txEvent)
    const multisigFindings = await app.MultisigWatcherSrv.handleTransaction(txEvent)

    const findings = [...vaultFindings, ...multisigFindings]
    app.healthChecker.check(findings)

    return findings
  }
}

export const healthCheck = (): HealthCheck => {
  return async function (): Promise<string[] | void> {
    const app = await App.getInstance()

    if (!app.healthChecker.isHealth()) {
      return ['There is too much network errors']
    }

    return []
  }
}

export default {
  initialize: initialize(),
  handleBlock: handleBlock(),
  handleTransaction: handleTransaction(),
  healthCheck: healthCheck(),
}
