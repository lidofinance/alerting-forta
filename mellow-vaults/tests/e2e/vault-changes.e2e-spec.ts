import { ethers, Finding, getEthersProvider, Network, Transaction } from 'forta-agent'
import { JsonRpcProvider } from '@ethersproject/providers'

import { App } from '../../src/app'
import { createTransactionEvent, etherBlockToFortaBlockEvent } from './utils'

const TEST_TIMEOUT = 60_000 // ms

describe('vault-acl-changes e2e tests', () => {
  let ethProvider: JsonRpcProvider
  // let getRoleMembers: (address: string, hash: string, currentBlock: BlockTag) => Promise<E.Either<Error, string[]>>
  let runTransaction: (txHash: string, initBlock?: number) => Promise<Finding[]>
  let runBlock: (blockHashOrNumber: string | number, initBlock?: number) => Promise<Finding[]>
  let logSpy: jest.SpyInstance
  let timeSpy: jest.SpyInstance

  beforeAll(() => {
    ethProvider = getEthersProvider()

    logSpy = jest.spyOn(console, 'log')
    logSpy.mockImplementation(() => {})
    timeSpy = jest.spyOn(Date, 'now')
    timeSpy.mockImplementation(() => new Date('2023-12-31'))

    runTransaction = async (txHash: string, initBlock?: number) => {
      const app = await App.getInstance()

      const receipt = await ethProvider.send('eth_getTransactionReceipt', [txHash])
      const block = await ethProvider.send('eth_getBlockByNumber', [
        ethers.utils.hexValue(parseInt(receipt.blockNumber)),
        true,
      ])
      const transaction = block.transactions.find((tx: Transaction) => tx.hash.toLowerCase() === txHash)!
      if (initBlock) {
        await app.VaultWatcherSrv.initialize(initBlock)
      }

      const txEvent = createTransactionEvent(transaction, block, Network.MAINNET, [], receipt.logs)

      return app.VaultWatcherSrv.handleTransaction(txEvent)
    }

    runBlock = async (blockHashOrNumber, initBlock) => {
      const app = await App.getInstance()

      const blockNumber = initBlock
      const block = await ethProvider.getBlock(blockHashOrNumber)
      if (blockNumber) {
        await app.VaultWatcherSrv.initialize(blockNumber)
      }
      const blockEvent = etherBlockToFortaBlockEvent(block)
      return app.VaultWatcherSrv.handleBlock(blockEvent)
    }
  })

  afterAll(() => {
    jest.resetAllMocks()
  })

  it(
    'should RoleRevokedRoleRevoked',
    async () => {
      const findings = await runTransaction('0xb8b96ee47cadb80ec41dafceba1fe5b10b1d50b75e7ad9114ee16180bda2d2b4')
      expect(findings).toMatchSnapshot()
    },
    TEST_TIMEOUT,
  )

  it(
    'should RoleGranted',
    async () => {
      const findings = await runTransaction('0x28cbf0521b4e6c03bf719cf958516609b3bf678b2f20dc958ae2db8999345fc5')
      expect(findings).toMatchSnapshot()
    },
    TEST_TIMEOUT,
  )

  it(
    'should RoleAdminChanged (3+2)',
    async () => {
      const findings = await runTransaction('0x047e8dd86f2e58fd05630e3fa1d1b1ce4a7638e3138cf9762894cf99a4fcabe6')
      expect(findings).toMatchSnapshot()
    },
    TEST_TIMEOUT,
  )

  it(
    'should be not slot changes',
    async () => {
      const findings = await runTransaction(
        '0xffef3e8046d882ccd5773582438ebf6576e4093dec734ed27e0f5065d1a89d7b',
        20061165,
      )
      expect(findings).toMatchSnapshot()
    },
    TEST_TIMEOUT,
  )

  it(
    'should everything is fine',
    async () => {
      const findings = await runBlock(20061165, 20061166)
      expect(findings).toMatchSnapshot()
    },
    TEST_TIMEOUT,
  )

  it(
    'should process tx with Added Owner and Execution Success',
    async () => {
      const findings = await runTransaction('0x4c3e507fd78daac24dd75e33e7c69381481db8bce62c795d2eec67126ad6d396')
      expect(findings).toMatchSnapshot()
    },
    TEST_TIMEOUT,
  )
  it(
    'should process tx with Changed Threshold and Execution Success',
    async () => {
      const findings = await runTransaction('0xb58e9e81ad1dac1f33b9dfc4d19f2d909a2a3ea890c31aa27a0df10f86bd4eea')
      expect(findings).toMatchSnapshot()
    },
    TEST_TIMEOUT,
  )
  it(
    'should withdrawal',
    async () => {
      const findings = await runTransaction('0x17c222325fa85abc5a0e74708264c751cbf54e6f06fdfd852d520e3bee1b0596')
      expect(findings).toMatchSnapshot()
    },
    TEST_TIMEOUT,
  )

  it(
    'should should find 1 withdrawal',
    async () => {
      const findings = await runBlock(20109999, 20109998)
      expect(findings).toMatchSnapshot()
    },
    TEST_TIMEOUT,
  )

  it(
    'should find no withdrawals at 2 vaults',
    async () => {
      const findings = await runBlock(20116800, 20116799)
      expect(findings).toMatchSnapshot()
    },
    TEST_TIMEOUT,
  )
})
