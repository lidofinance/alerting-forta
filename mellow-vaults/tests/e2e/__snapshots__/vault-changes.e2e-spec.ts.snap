// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`vault-acl-changes e2e tests should RoleAdminChanged (3+2) 1`] = `
[
  Finding {
    "addresses": [],
    "alertId": "MELLOW-VAULT-ROLE-ADMIN-CHANGED",
    "description": "Role Admin for role 0xf23ec0bb4210edd5cba85afd05127efcd2fc6a781bfed49188da1081670b22d8(ADMIN ROLE) was changed from 0x0000000000000000000000000000000000000000000000000000000000000000 to 0xf23ec0bb4210edd5cba85afd05127efcd2fc6a781bfed49188da1081670b22d8 bond strategy address - 0x7a14b34a9a8ea235c66528dc3bf3aefc36dfc268",
    "labels": [],
    "metadata": {
      "args": "0xf23ec0bb4210edd5cba85afd05127efcd2fc6a781bfed49188da1081670b22d8,0x0000000000000000000000000000000000000000000000000000000000000000,0xf23ec0bb4210edd5cba85afd05127efcd2fc6a781bfed49188da1081670b22d8",
    },
    "name": "üö® Vault: Role Admin changed",
    "protocol": "ethereum",
    "severity": 4,
    "source": {},
    "timestamp": 2023-12-31T00:00:00.000Z,
    "type": 4,
    "uniqueKey": "",
  },
  Finding {
    "addresses": [],
    "alertId": "MELLOW-VAULT-ROLE-ADMIN-CHANGED",
    "description": "Role Admin for role 0xc171260023d22a25a00a2789664c9334017843b831138c8ef03cc8897e5873d7(ADMIN DELEGATE ROLE) was changed from 0x0000000000000000000000000000000000000000000000000000000000000000 to 0xf23ec0bb4210edd5cba85afd05127efcd2fc6a781bfed49188da1081670b22d8 bond strategy address - 0x7a14b34a9a8ea235c66528dc3bf3aefc36dfc268",
    "labels": [],
    "metadata": {
      "args": "0xc171260023d22a25a00a2789664c9334017843b831138c8ef03cc8897e5873d7,0x0000000000000000000000000000000000000000000000000000000000000000,0xf23ec0bb4210edd5cba85afd05127efcd2fc6a781bfed49188da1081670b22d8",
    },
    "name": "üö® Vault: Role Admin changed",
    "protocol": "ethereum",
    "severity": 4,
    "source": {},
    "timestamp": 2023-12-31T00:00:00.000Z,
    "type": 4,
    "uniqueKey": "",
  },
  Finding {
    "addresses": [],
    "alertId": "MELLOW-VAULT-ROLE-ADMIN-CHANGED",
    "description": "Role Admin for role 0x46a52cf33029de9f84853745a87af28464c80bf0346df1b32e205fc73319f622(OPERATOR) was changed from 0x0000000000000000000000000000000000000000000000000000000000000000 to 0xc171260023d22a25a00a2789664c9334017843b831138c8ef03cc8897e5873d7 bond strategy address - 0x7a14b34a9a8ea235c66528dc3bf3aefc36dfc268",
    "labels": [],
    "metadata": {
      "args": "0x46a52cf33029de9f84853745a87af28464c80bf0346df1b32e205fc73319f622,0x0000000000000000000000000000000000000000000000000000000000000000,0xc171260023d22a25a00a2789664c9334017843b831138c8ef03cc8897e5873d7",
    },
    "name": "üö® Vault: Role Admin changed",
    "protocol": "ethereum",
    "severity": 4,
    "source": {},
    "timestamp": 2023-12-31T00:00:00.000Z,
    "type": 4,
    "uniqueKey": "",
  },
  Finding {
    "addresses": [],
    "alertId": "MELLOW-VAULT-ROLE-GRANTED",
    "description": "Role 0x46a52cf33029de9f84853745a87af28464c80bf0346df1b32e205fc73319f622(OPERATOR) was granted to 0x188858AC61a74350116d1CB6958fBc509FD6afA1 by 0x188858AC61a74350116d1CB6958fBc509FD6afA1 bond strategy address - 0x7a14b34a9a8ea235c66528dc3bf3aefc36dfc268",
    "labels": [],
    "metadata": {
      "args": "0x46a52cf33029de9f84853745a87af28464c80bf0346df1b32e205fc73319f622,0x188858AC61a74350116d1CB6958fBc509FD6afA1,0x188858AC61a74350116d1CB6958fBc509FD6afA1",
    },
    "name": "üö® Vault: Role granted",
    "protocol": "ethereum",
    "severity": 4,
    "source": {},
    "timestamp": 2023-12-31T00:00:00.000Z,
    "type": 4,
    "uniqueKey": "",
  },
  Finding {
    "addresses": [],
    "alertId": "MELLOW-VAULT-ROLE-GRANTED",
    "description": "Role 0xf23ec0bb4210edd5cba85afd05127efcd2fc6a781bfed49188da1081670b22d8(ADMIN ROLE) was granted to 0x188858AC61a74350116d1CB6958fBc509FD6afA1 by 0x188858AC61a74350116d1CB6958fBc509FD6afA1 bond strategy address - 0x7a14b34a9a8ea235c66528dc3bf3aefc36dfc268",
    "labels": [],
    "metadata": {
      "args": "0xf23ec0bb4210edd5cba85afd05127efcd2fc6a781bfed49188da1081670b22d8,0x188858AC61a74350116d1CB6958fBc509FD6afA1,0x188858AC61a74350116d1CB6958fBc509FD6afA1",
    },
    "name": "üö® Vault: Role granted",
    "protocol": "ethereum",
    "severity": 4,
    "source": {},
    "timestamp": 2023-12-31T00:00:00.000Z,
    "type": 4,
    "uniqueKey": "",
  },
]
`;

exports[`vault-acl-changes e2e tests should RoleGranted 1`] = `
[
  Finding {
    "addresses": [],
    "alertId": "MELLOW-VAULT-ROLE-GRANTED",
    "description": "Role 0x46a52cf33029de9f84853745a87af28464c80bf0346df1b32e205fc73319f622(OPERATOR) was granted to 0x2afc096981c2CFe3501bE4054160048718F6C0C8 by 0x188858AC61a74350116d1CB6958fBc509FD6afA1 bond strategy address - 0x7a14b34a9a8ea235c66528dc3bf3aefc36dfc268",
    "labels": [],
    "metadata": {
      "args": "0x46a52cf33029de9f84853745a87af28464c80bf0346df1b32e205fc73319f622,0x2afc096981c2CFe3501bE4054160048718F6C0C8,0x188858AC61a74350116d1CB6958fBc509FD6afA1",
    },
    "name": "üö® Vault: Role granted",
    "protocol": "ethereum",
    "severity": 4,
    "source": {},
    "timestamp": 2023-12-31T00:00:00.000Z,
    "type": 4,
    "uniqueKey": "",
  },
]
`;

exports[`vault-acl-changes e2e tests should RoleRevokedRoleRevoked 1`] = `
[
  Finding {
    "addresses": [],
    "alertId": "MELLOW-VAULT-ROLE-REVOKED",
    "description": "Role 0xc171260023d22a25a00a2789664c9334017843b831138c8ef03cc8897e5873d7(ADMIN DELEGATE ROLE) was revoked to 0x188858AC61a74350116d1CB6958fBc509FD6afA1 by 0x188858AC61a74350116d1CB6958fBc509FD6afA1 bond strategy address - 0x7a14b34a9a8ea235c66528dc3bf3aefc36dfc268",
    "labels": [],
    "metadata": {
      "args": "0xc171260023d22a25a00a2789664c9334017843b831138c8ef03cc8897e5873d7,0x188858AC61a74350116d1CB6958fBc509FD6afA1,0x188858AC61a74350116d1CB6958fBc509FD6afA1",
    },
    "name": "üö®  Vault: Role revoked",
    "protocol": "ethereum",
    "severity": 4,
    "source": {},
    "timestamp": 2023-12-31T00:00:00.000Z,
    "type": 4,
    "uniqueKey": "",
  },
]
`;

exports[`vault-acl-changes e2e tests should be not slot changes 1`] = `[]`;

exports[`vault-acl-changes e2e tests should everything is fine 1`] = `[]`;

exports[`vault-acl-changes e2e tests should process tx with Added Owner and Execution Success 1`] = `[]`;

exports[`vault-acl-changes e2e tests should process tx with Changed Threshold and Execution Success 1`] = `[]`;

exports[`vault-acl-changes e2e tests should should find 1 withdrawal 1`] = `[]`;

exports[`vault-acl-changes e2e tests should find no withdrawals at 2 vaults 1`] = `
[
  Finding {
    "addresses": [],
    "alertId": "VAULT-LIMITS-INTEGRITY-CLOSE-TO-MAX",
    "description": "Mellow Vault 0xbeef69ac7870777598a04b2bd4771c71212e6abc totalSupply more than 90% of maximalTotalSupply",
    "labels": [],
    "metadata": {},
    "name": "‚ö†Ô∏è Vault totalSupply close to maximalTotalSupply",
    "protocol": "ethereum",
    "severity": 3,
    "source": {},
    "timestamp": 2023-12-31T00:00:00.000Z,
    "type": 2,
    "uniqueKey": "",
  },
  Finding {
    "addresses": [],
    "alertId": "VAULT-LIMITS-INTEGRITY-REACHED-TO-MAX",
    "description": "Mellow Vault 0x7a4effd87c2f3c55ca251080b1343b605f327e3a maximalTotalSupply reached",
    "labels": [],
    "metadata": {},
    "name": "‚ö†Ô∏è Vault totalSupply reached maximalTotalSupply",
    "protocol": "ethereum",
    "severity": 3,
    "source": {},
    "timestamp": 2023-12-31T00:00:00.000Z,
    "type": 2,
    "uniqueKey": "",
  },
  Finding {
    "addresses": [],
    "alertId": "VAULT-WSTETH-LIMIT-REACHED",
    "description": "Symbiotic - 0xc329400492c6ff2438472d4651ad17389fcb843a",
    "labels": [],
    "metadata": {},
    "name": "‚ö†Ô∏è Vault: Symbiotic limit reached",
    "protocol": "ethereum",
    "severity": 3,
    "source": {},
    "timestamp": 2023-12-31T00:00:00.000Z,
    "type": 2,
    "uniqueKey": "",
  },
  Finding {
    "addresses": [],
    "alertId": "VAULT-NO-WITHDRAWAL-48",
    "description": "Mellow Vault - 0x84631c0d0081fde56deb72f6de77abbbf6a9f93a",
    "labels": [],
    "metadata": {},
    "name": "‚ö†Ô∏è Vault: Withdrawals haven‚Äôt been called for at least 48 hours",
    "protocol": "ethereum",
    "severity": 3,
    "source": {},
    "timestamp": 2023-12-31T00:00:00.000Z,
    "type": 2,
    "uniqueKey": "",
  },
  Finding {
    "addresses": [],
    "alertId": "VAULT-NO-WITHDRAWAL-48",
    "description": "Mellow Vault - 0x7a4effd87c2f3c55ca251080b1343b605f327e3a",
    "labels": [],
    "metadata": {},
    "name": "‚ö†Ô∏è Vault: Withdrawals haven‚Äôt been called for at least 48 hours",
    "protocol": "ethereum",
    "severity": 3,
    "source": {},
    "timestamp": 2023-12-31T00:00:00.000Z,
    "type": 2,
    "uniqueKey": "",
  },
]
`;

exports[`vault-acl-changes e2e tests should withdrawal 1`] = `
[
  Finding {
    "addresses": [],
    "alertId": "MELLOW-VAULT-WITHDRAWAL-ALL",
    "description": "Vault withdrawal, bond strategy address - 0x7a14b34a9a8ea235c66528dc3bf3aefc36dfc268",
    "labels": [],
    "metadata": {
      "args": "0xE9229230D3768b42e7c2a54cffbfE0f0288261aC,0x7fF11ad66Ec323C6A5D40363946De553D8cf2B2E,0x1A63572CF23505DDBCCB2Bdd0cE516Be9f98E842,0x21ded9833e786Ea941410312e7e78Fc79fc879Af,0x9A3b9ae785C399bcdb04115fe78ad866EE39c877,1718609063",
    },
    "name": "‚ÑπÔ∏è Vault: Withdrawal all",
    "protocol": "ethereum",
    "severity": 1,
    "source": {},
    "timestamp": 2023-12-31T00:00:00.000Z,
    "type": 4,
    "uniqueKey": "",
  },
]
`;
